lib/RestructureCFG/InlineDispatcherSwitch.cpp

```diff
@@ -737,13 +737,140 @@ inlineDispatcherSwitchImpl(ASTTree &AST,
+    revng_assert(not Seq->isEmpty());
+  } break;
+  case ASTNode::NK_Scs: {
+    ScsNode *Scs = llvm::cast<ScsNode>(Node);
+
+    // Inspect loop nodes
+    if (Scs->hasBody()) {
+      ASTNode *Body = Scs->getBody();
+      ASTNode *NewBody = rc_recur
+        simplifySuperfluousNonLocalCFImpl(Body, FallThroughScopeMap);
+      Scs->setBody(NewBody);
+    }
+
+  } break;
+  case ASTNode::NK_If: {
+    IfNode *If = llvm::cast<IfNode>(Node);
+
+    // Inspect the `then` and `else` branches
+    if (If->hasThen()) {
+      ASTNode *Then = If->getThen();
+      ASTNode *NewThen = rc_recur
+        simplifySuperfluousNonLocalCFImpl(Then, FallThroughScopeMap);
+      If->setThen(NewThen);
+    }
+    if (If->hasElse()) {
+      ASTNode *Else = If->getElse();
+      ASTNode *NewElse = rc_recur
+        simplifySuperfluousNonLocalCFImpl(Else, FallThroughScopeMap);
+      If->setElse(NewElse);
+    }
+  } break;
+  case ASTNode::NK_Switch: {
+    auto *Switch = llvm::cast<SwitchNode>(Node);
+
+    // First of all, we recursively process the `case` nodes contained in the
+    // `switch` in order to process the inner portion of the AST
+    llvm::SmallVector<size_t> ToRemoveCaseIndex;
+    for (auto &Group : llvm::enumerate(Switch->cases())) {
+      unsigned Index = Group.index();
+      auto &LabelCasePair = Group.value();
+      LabelCasePair.second = rc_recur
+        simplifySuperfluousNonLocalCFImpl(LabelCasePair.second,
+                                          FallThroughScopeMap);
+
+      if (LabelCasePair.second == nullptr) {
+        ToRemoveCaseIndex.push_back(Index);
+      }
+    }
+
+    for (auto ToRemoveCase : llvm::reverse(ToRemoveCaseIndex)) {
+      Switch->removeCaseN(ToRemoveCase);
+    }
+  } break;
+  case ASTNode::NK_Code:
+  case ASTNode::NK_Set:
+  case ASTNode::NK_SwitchBreak:
+  case ASTNode::NK_Continue:
+  case ASTNode::NK_Break:
+    // Do nothing
+    break;
+  default:
+    revng_unreachable();
+  }
+
+  rc_return Node;
+}
+
+/// This simplification routine is aimed at removing superfluous
+/// `NonLocalControl Flow` statements that may remain around after the inline of
+/// `cases`. Specifically, if a `return` statement is moved into an inner loop
+/// in place of a `SetNode`, it may be that a following `break` statement can be
+/// simplified away.
+static ASTNode *simplifySuperfluousNonLocalCF(ASTTree &AST,
+                                              ASTNode *RootNode,
+                                              const FallThroughScopeTypeMap
+                                                &FallThroughScopeMap) {
+
+  // Call the underlying implementation
+  RootNode = simplifySuperfluousNonLocalCFImpl(RootNode, FallThroughScopeMap);
+  AST.setRoot(RootNode);
+
+  return RootNode;
+}
+
+ASTNode *inlineDispatcherSwitch(const model::Binary &Model,
+                                ASTTree &AST,
+                                ASTNode *RootNode) {
```

Comment:
At this point I don't think you even need to take the `RootNode` argument explicitly.
It seems to me that the whole code always assumes that `RootNode == AST.getRoot()`. Isn't that the case?
If I'm right, passing `RootNode` explicitly opens up opportunities for callers to pass a wrong `RootNode`, so I'd just remove it.

Same holds for `inlineDispatcherSwitch` below. Please fix that too.

Reply:
I crafted a new commit dedicated to removing all the unnecessary `RootNode` argument b9a8a08f7f6160f57da4506aea7fce6f85550453